### PR TITLE
fix(txpool): reject transaction with malformed 'to' at admission time (#5318)

### LINKED ISSUE #5318: [Bug] A single transaction with a non-hex `to` permanently halts consensus on v3.16.x (BASELINE_SCHEDULER boost unhex throw + PBFT infinite re-propose)
**Describe the bug**

A single transaction whose `to` field is not a valid hex string permanently halts consensus on v3.16.x. The `BASELINE_SCHEDULER` throws while hex-decoding `to` during block execution; PBFT treats the deterministic execution failure as transient and re-proposes the *same* block forever. Block height stops advancing on every node, the pending transaction that triggered it (plus everything after) never gets a receipt, and the chain does not recover on its own.

This is a consensus-liveness / robustness issue: an invalid `to` passes txpool verification, gets packed into a block, and then deterministically fails execution — which consensus cannot make progress past. On v3.7.3 the same transaction is tolerated (the older executor path does not throw), so only recent releases are affected.

**To Reproduce**

1. Start a 4-node air chain from the released `v3.16.3` binary (`build_chain.sh -p 30300,20200 -l 127.0.0.1:4`).
2. Submit **one** transaction whose `to` field is a non-hex string — e.g. the raw ASCII bytes of `"HelloWorld"` instead of a 20-byte hex address / empty string. (In our case this came from the Java SDK: `AssembleTransactionProcessor.deployAndGetResponseWithStringParams(abi, bin, params, path)` copies its `path` argument verbatim into the transaction `to` field via `createSignedTransaction(...)`; on a Solidity chain `path` must be `""`, but any client that puts a non-hex value in `to` reproduces it.)
3. Observe every node stop producing blocks; all subsequent transactions time out with no receipt.

**Node log signature** (from the leader executing the poisoned block):

```
warning|[CONSENSUS][Core]asyncExecuteBlock failed,number=575,code=-70000,message=Execute block failed! .../boost/algorithm/hex.hpp(76): Throw in function unsigned char boost::algorithm::detail::hex_char_to_int(T) [T = char]
warning|[CONSENSUS][PBFT]proposal execute failed,propHash=257a94c1...,propIndex=575,committedIndex=574,consNum=575,view=6,toView=6,...
error|[BASELINE_SCHEDULER]Execute block failed! .../boost/algorithm/hex.hpp(76): Throw in function unsigned char boost::algorithm::detail::hex_char_to_int(T) [T = char]
```

These three lines then repeat continuously (the same `propIndex=575` / `committedIndex=574` is retried without end).

**Chain-frozen evidence** — height and txpool identical on all 4 nodes and unchanged over 5s while the poisoned proposal loops:

```
[node0 :20210] blockNumber: 574   pendingTxSize: 44
[node1 :20211] blockNumber: 574   pendingTxSize: 44
[node2 :20212] blockNumber: 574   pendingTxSize: 44
[node3 :20213] blockNumber: 574   pendingTxSize: 44
   ... (blockNumber still 574 after 5s) ...
```

**Expected behavior**

A transaction with a malformed `to` (not valid hex, or wrong length) should be rejected at txpool / transaction-verification time and never packed into a block. Failing that, the executor must not throw an unhandled exception on such input, and — more importantly — consensus must not livelock re-proposing a block that deterministically fails execution: a permanently un-executable proposal should be dropped/rolled back so the chain keeps making progress, rather than one bad transaction halting the whole network.

**Environment**

- OS: reproduced on both Ubuntu (`ubuntu-latest`, x86_64) and macOS (`macos-latest`, x86_64) GitHub-hosted runners.
- FISCO BCOS Version: **v3.16.3** affected (BASELINE_SCHEDULER). **v3.7.3 not affected** (same transaction is tolerated).
- Deployment: local 4-node air chain via `build_chain.sh`, RPC SSL disabled, default (P)BFT config.

**Additional context**

- Discovered while stabilizing the java-sdk integration-test CI (FISCO-BCOS/java-sdk PR #961). On the SDK side we already stopped sending the malformed `to`; this report is about the **node-side** halt, since any client able to submit such a transaction can stop a v3.16.x chain.
- The impact is availability, not just a single failed transaction: the chain does not self-heal and every later transaction on the group is stuck until the poisoned proposal is somehow cleared.

### PR DESCRIPTION
**Pull request by kyonRay**
**Motivation**

Fixes #5318. A single transaction whose `to` field is not a valid hex address permanently halts consensus on v3.16.x: it passes txpool verification, gets packed into a block, then `BASELINE_SCHEDULER` throws inside `boost::algorithm::unhex` while decoding `to` (`unhexAddress`, called from `newEVMCMessage`). The scheduler maps the exception to `SchedulerError::UnknownError`, and PBFT treats the deterministic failure as transient, re-proposing the same block forever. Block height stops advancing on every node and the chain never recovers.

**Solution**

Validate the `to` field at txpool admission time: it must be empty (deployment) or a 20-byte hex address with an optional `0x`/`0X` prefix (`isValidToField()`). The check is enforced on both paths a transaction can enter the pool through:

1. `TxValidator::validateTransaction()` — client submission via RPC and p2p tx sync is rejected with `TransactionStatus::Malformed`, so a malformed `to` can never be sealed by an honest node.
2. `MemoryStorage::enforceSubmitTransaction()` — transactions missing from the local pool during proposal verification are imported through this path, which bypasses `validateTransaction()`. Re-checking here makes a proposal that carries a malformed `to` (from a malicious or unfixed leader) fail verification, so PBFT view-changes to a leader with a clean proposal instead of executing the poisoned block.

WASM chains are exempt (`g_BCOSConfig.isWasm()`) because their `to` carries a BFS path (e.g. `/apps/HelloWorld`) instead of an address. Web3 transactions are unaffected: their `to` is always written as `hexPrefixed()` of a 20-byte RLP address. Admission-level rejection changes no execution semantics for any transaction already accepted on-chain, so there is no hardfork risk.

Regression test `Issue5318InvalidToTest.cpp` covers the original repro (`"HelloWorld"` as `to`), 40-char non-hex, wrong lengths, prefix edge cases, valid address forms, the web3 type, the WASM exemption, and the enforce-import path (positive control: a valid transaction imports through the same path; poison transaction is rejected).

### CODE REVIEW

**Review by ywy2090 — APPROVED** [review]

File: bcos-txpool/bcos-txpool/txpool/validator/TxValidator.cpp [hunk 1]
```diff
@@ -142,6 +160,14 @@ bcos::protocol::TransactionStatus TxValidator::checkWeb3Nonce(
 
 TransactionStatus TxValidator::validateTransaction(const bcos::protocol::Transaction& _tx)
 {
+    // Issue #5318: reject a malformed `to` at admission time so it can never reach a block.
+    if (!isValidToField(_tx.to()))
+    {
+        TX_VALIDATOR_CHECKER_LOG(WARNING)
+            << LOG_BADGE("ValidateTransaction") << LOG_DESC("RejectTransactionWithInvalidTo")
+            << LOG_KV("to", _tx.to()) << LOG_KV("hash", _tx.hash().abridged());
```
Comment (morebtcg) [inline bcos-txpool/bcos-txpool/txpool/validator/TxValidator.cpp:168]: **Log DoS risk**: `LOG_KV("to", _tx.to())` logs the raw `to` field without a size limit. A crafted transaction with a very large `to` field could cause excessive log volume / memory pressure when rejected. Consider logging a truncated value (e.g. first 64 chars) plus the original length instead:

```cpp
<< LOG_KV("to", _tx.to().substr(0, 64)) << LOG_KV("toSize", _tx.to().size())
```

File: bcos-txpool/bcos-txpool/txpool/storage/MemoryStorage.cpp [hunk 2]
```diff
@@ -274,6 +274,19 @@ TransactionStatus MemoryStorage::txpoolStorageCheck(
...
     auto txHash = _tx->hash();
+    // Issue #5318: transactions on this path come from another node's proposal and bypass
+    // validateTransaction(), so re-check `to` here — otherwise a proposal carrying a
+    // malformed `to` is imported, passes verification and deterministically fails
+    // execution, halting consensus. Rejecting it fails the proposal verification instead,
+    // and PBFT view-changes to a leader with a clean proposal.
+    if (!isValidToField(_tx->to()))
+    {
+        TXPOOL_LOG(WARNING) << LOG_DESC("enforce to seal failed for malformed to field")
+                            << LOG_KV("to", _tx->to()) << LOG_KV("importTxHash", txHash.abridged())
```
Comment (morebtcg) [inline bcos-txpool/bcos-txpool/txpool/storage/MemoryStorage.cpp:285]: **Same Log DoS risk as above**: `LOG_KV("to", _tx->to())` logs the raw `to` field on the enforce-import path as well. Please apply the same truncation treatment for consistency.

File: bcos-txpool/test/unittests/txpool/Issue5318InvalidToTest.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    txpoolStorage->clear();
+}
+
+BOOST_AUTO_TEST_CASE(wasmChainSkipsAddressCheck)
+{
+    auto validator = makeValidator();
+
+    // on a WASM chain `to` is a BFS path, not a hex address — the check must not apply
+    g_BCOSConfig.setIsWasm(true);
```
Comment (morebtcg) [inline bcos-txpool/test/unittests/txpool/Issue5318InvalidToTest.cpp:146]: **Test state leakage**: This test mutates `g_BCOSConfig.setIsWasm(true)` and `setIsWasm(false)` without saving/restoring the original value. If the test suite is ever run in an environment where `isWasm` is already `true`, this would leak `false` to subsequent tests. Consider:

```cpp
auto const oldIsWasm = g_BCOSConfig.isWasm();
g_BCOSConfig.setIsWasm(true);
// ... WASM tests ...
g_BCOSConfig.setIsWasm(false);
// ... non-WASM tests ...
g_BCOSConfig.setIsWasm(oldIsWasm);
```

**Review by morebtcg — COMMENTED** [review]
Overall this is a well-designed fix that correctly addresses the consensus-liveness vulnerability in #5318. The dual-path enforcement (RPC submission + proposal-import) is thorough and the WASM exemption is correctly placed. The test coverage is comprehensive. Below are a few non-blocking suggestions for improvement.

**Review by morebtcg — APPROVED** [review]
LGTM. No blocking issues found. The fix correctly addresses the consensus-liveness vulnerability with thorough dual-path enforcement and comprehensive test coverage. The 3 inline suggestions (log truncation, test state save/restore) are non-blocking improvements.
